README.md
# `vscode-redmine`

Redmine extension for Visual Studio Code.

## Features

- Sidebar panel
  - List issues assigned to you in sidebar panel
  - List projects and open issues in them
- Create issue (opens redmine create issue in browser)
- List of issues assigned to you
- Open issue by id
- Open issue by selected number in document
- Issue actions:
  - Change status of an issue
  - Add time entry to an issue
  - Open issue in browser
  - Quick update issue

_Missing a feature? Open an issue and let me know!_

### Sidebar panel

![Sidebar panel GIF showcase](./resources/redmine-sidebar-panel.gif)

### Add time entry from action menu

![Add time entry from action menu GIF showcase](./resources/redmine-add-time-entry.gif)

### Change server to other workspace folder in sidebar panel

![Change server to other workspace folder in sidebar panel GIF showcase](./resources/redmine-change-sidebar-server.gif)

## Requirements

It's required to enable REST web services in `/settings?tab=api` of your redmine (you have to be administrator of redmine server).

## Extension Settings

This extension contributes the following settings:

- `redmine.url`: URL of redmine server (eg. `https://example.com`, `http://example.com:8080`, `https://example.com:8443/redmine`, `http://example.com/redmine` _etc._)
- `redmine.apiKey`: API Key of your redmine account (see `/my/account` page, on right-hand pane)
- `redmine.rejectUnauthorized`: Parameter, which is passed to https request options (true/false) (useful to fix issues with self-signed certificates, see issue #3)
- `redmine.identifier`: If set, this will be the project, to which new issue will be created

  _NOTE: this is an identifier of project, not display name of the project_

- `redmine.additionalHeaders`: Object of additional headers to be sent along with every request to redmine server

## Contribution

If you want to contribute to the project, please read [contributing guide](./CONTRIBUTING.md) guide.

## Known Issues

No known issues yet. If you found one, feel free to open an issue!

## Release Notes

See [change log](./CHANGELOG.md)

## Attributions

### Logo

Logo is remixed version of original Redmine Logo.

Redmine Logo is Copyright (C) 2009 Martin Herr and is licensed under the Creative Commons Attribution-Share Alike 2.5 Generic license.
See http://creativecommons.org/licenses/by-sa/2.5/ for more details.

Made with ❤️ by [@rozpuszczalny](https://tomaszdomanski.com/?utm_source=github&utm_campaign=vscode-redmine)


CHANGELOG.md
# Change Log

All notable changes to the "vscode-redmine" extension will be documented in this file.

## [Unreleased]

## 1.1.1 - 17.08.2022

### Fixed

- [Issue #53](https://github.com/rozpuszczalny/vscode-redmine/issues/53)

## 1.1.0 - 19.01.2022

### Added

- [Issue #34](https://github.com/rozpuszczalny/vscode-redmine/issues/34) - display subprojects as a tree and toggle back to flat view (thanks to [doganOz](https://github.com/doganOz)!)
- Code of conduct, contributing guide and issue template

### Changed

- eslint rules to be more restrictive
- small refactor

## 1.0.4 - 13.10.2021

### Fixed

- [Issue #37](https://github.com/rozpuszczalny/vscode-redmine/issues/37)

## 1.0.3 - 21.09.2020

### Fixed

- [Issue #30](https://github.com/rozpuszczalny/vscode-redmine/issues/30)

## 1.0.2 - 25.08.2020

### Added

- Updated link 😇

## 1.0.1 - 25.08.2020

### Added

- Link to my website

## 1.0.0 - 25.08.2020

### Fixed

- [Issue #23](https://github.com/rozpuszczalny/vscode-redmine/issues/23)

### Added

- [Issue #18](https://github.com/rozpuszczalny/vscode-redmine/issues/18) Showcase of features in README.md
- [Issue #16](https://github.com/rozpuszczalny/vscode-redmine/issues/16) Mulitroot support - configuration can be set in `settings.json` file, under `.vscode` folder of workspace folder.
- ESLint and Prettier
- [Issue #11](https://github.com/rozpuszczalny/vscode-redmine/issues/11) Sidebar panel with issues assigned to user and list of projects

### Removed

- 💥 `serverUrl`, `serverPort`, `serverIsSsl` and `authorization` settings; use `url` and `additionalHeaders` instead.

### Changed

- Bumped to TypeScript 3
- Align code to ESLint suggestions
- Format codebase with Prettier

## 0.5.1 - 30.01.2019

### Changed

- Updated README.md with new features description

## 0.5.0 - 30.01.2019

### Added

Thanks to **[MarkusAmshove](https://github.com/MarkusAmshove)**:

- Quick update action for issue
  - You can quickly change issue status, assigned and add note
- Open issue actions from number selected in document

## 0.4.0 - 16.09.2018

### Added

- Custom `Authorization` header
  - You can configure `Authorization` header by setting `redmine.authorization`.

## 0.3.1 - 20.08.2018

### Fixed

- [Issue #7](https://github.com/rozpuszczalny/vscode-redmine/issues/7) (thanks to **[hanyuzhou2006](https://github.com/hanyuzhou2006)** for discovering and fixing an issue!)

## 0.3.0 - 21.06.2018

### Added

- Create issue command
  - You can configure project name under `redmine.projectName` setting
  - You can choose project from the list, if `redmine.projectName` isn't configured
- 0.2.2 `CHANGELOG.md` entry

### Changed

- 0.2.1 correct year in `CHANGELOG.md`

## 0.2.2 - 07.06.2018

### Added

- Attribution of Redmine Logo

## 0.2.1 - 07.06.2018

### Added

- Extension icon

### Changed

- Display name

## 0.2.0 - 02.06.2018

### Added

- `rejectUnauthorized` parameter, which is passed to https request options
- Missing open by issue id in feature list in `README.md`
- Information about future features in `README.md`

## 0.1.1 - 14.03.2018

### Fixed

- messy issue description in quick pick

## 0.1.0 - 04.02.2018

Added more basic functionalities to this extensions.

### Added

- possibility to change issue status
- possibility to add time entries to issue
- getting issue actions by typing in an issue id

### Changed

- splitted changelog entries into separate sections

### Removed

- `Release notes` section in `README.md` - it can be viewed here, so there is no reason to copy that across multiple files

## 0.0.2 - 28.01.2018

### Added

- repository URL to `package.json`, so VSCode Marketplace see `README.md` and other files

## 0.0.1 - 28.01.2018

Initial release of `vscode-redmine`

### Added

- list of issues assigned to API key owner
- possibility to open issue in browser
- configuration for server and API key


src/redmine/redmine-server.ts
import { Url, parse } from "url";
import * as http from "http";
import * as https from "https";
import { RedmineProject } from "./redmine-project";
import {
  IssueStatus,
  Membership,
  QuickUpdate,
  QuickUpdateResult,
} from "../controllers/domain";
import { TimeEntryActivity } from "./models/time-entry-activity";
import { Project } from "./models/project";
import { TimeEntry } from "./models/time-entry";
import { Issue } from "./models/issue";
import { IssueStatus as RedmineIssueStatus } from "./models/issue-status";
import { Membership as RedmineMembership } from "./models/membership";
import isNil from "lodash/isNil";
import isEqual from "lodash/isEqual";

type HttpMethods = "GET" | "POST" | "PUT" | "DELETE" | "PATCH";

const REDMINE_API_KEY_HEADER_NAME = "X-Redmine-API-Key";

export interface RedmineServerConnectionOptions {
  /**
   * @example https://example.com
   * @example http://example.com:8080
   * @example https://example.com:8443/redmine
   * @example http://example.com/redmine
   */
  address: string;
  /**
   * @example 7215ee9c7d9dc229d2921a40e899ec5f
   */
  key: string;
  /**
   * @default false
   */
  rejectUnauthorized?: boolean;
  /**
   * @example { "Authorization": "Basic YTph" }
   */
  additionalHeaders?: { [key: string]: string };
}

interface RedmineServerOptions extends RedmineServerConnectionOptions {
  url: Url;
}

export class RedmineOptionsError extends Error {
  name = "RedmineOptionsError";
}

export class RedmineServer {
  options!: RedmineServerOptions;

  private timeEntryActivities: TimeEntryActivity[] | null = null;

  get request() {
    return this.options.url.protocol === "https:"
      ? https.request
      : http.request;
  }

  private validateOptions(options: RedmineServerConnectionOptions): void {
    if (!options.address) {
      throw new RedmineOptionsError("Address cannot be empty!");
    }
    if (!options.key) {
      throw new RedmineOptionsError("Key cannot be empty!");
    }
    const url = parse(options.address);
    if (["https:", "http:"].indexOf(url.protocol ?? "") === -1) {
      throw new RedmineOptionsError(
        "Address must have supported protocol (http/https)"
      );
    }
  }

  private setOptions(options: RedmineServerConnectionOptions) {
    this.options = {
      ...options,
      url: parse(options.address),
    };
    if (isNil(this.options.additionalHeaders)) {
      this.options.additionalHeaders = {};
    }
  }

  constructor(options: RedmineServerConnectionOptions) {
    this.validateOptions(options);
    this.setOptions(options);
  }

  doRequest<T>(path: string, method: HttpMethods, data?: Buffer): Promise<T> {
    const { url, key, additionalHeaders, rejectUnauthorized } = this.options;
    const options: https.RequestOptions = {
      hostname: url.hostname,
      port: url.port,
      headers: {
        [REDMINE_API_KEY_HEADER_NAME]: key,
        ...additionalHeaders,
      },
      rejectUnauthorized: rejectUnauthorized,
      path: `${url.pathname}${path}`,
      method,
    };
    if (data) {
      options.headers!["Content-Length"] = data.length;
      options.headers!["Content-Type"] = "application/json";
    }

    return new Promise((resolve, reject) => {
      let incomingBuffer = Buffer.from("");
      const handleData = (_: http.IncomingMessage) => (incoming: Buffer) => {
        incomingBuffer = Buffer.concat([incomingBuffer, incoming]);
      };

      const handleEnd = (clientResponse: http.IncomingMessage) => () => {
        const { statusCode, statusMessage } = clientResponse;
        if (statusCode === 401) {
          reject(
            new Error(
              "Server returned 401 (perhaps your API Key is not valid, or your server has additional authentication methods?)"
            )
          );
          return;
        }
        if (statusCode === 403) {
          reject(
            new Error(
              "Server returned 403 (perhaps you haven't got permissions?)"
            )
          );
          return;
        }
        if (statusCode === 404) {
          reject(new Error("Resource doesn't exist"));
          return;
        }

        // TODO: Other errors handle
        if (statusCode! >= 400) {
          reject(new Error(`Server returned ${statusMessage}`));
          return;
        }

        if (incomingBuffer.length > 0) {
          try {
            const object = JSON.parse(incomingBuffer.toString("utf8"));
            resolve(object);
          } catch (e) {
            reject(new Error("Couldn't parse Redmine response as JSON..."));
          }
          return;
        }

        // Using `doRequest` on the endpoints that return 204 should type as void/null
        resolve((null as unknown) as T);
      };

      const clientRequest = this.request(options, (incoming) => {
        incoming.on("data", handleData(incoming));
        incoming.on("end", handleEnd(incoming));
      });

      const handleError = (error: Error) => {
        reject(
          new Error(`NodeJS Request Error (${error.name}): ${error.message}`)
        );
      };

      clientRequest.on("error", handleError);

      clientRequest.end(data);
    });
  }

  async getProjects(): Promise<RedmineProject[]> {
    const req = async (
      offset = 0,
      limit = 50,
      count: number | null = null,
      accumulator: RedmineProject[] = []
    ): Promise<RedmineProject[]> => {
      if (count && count <= offset) {
        return accumulator;
      }

      const [totalCount, result]: [
        number,
        RedmineProject[]
      ] = await this.doRequest<{
        projects: Project[];
        total_count: number;
      }>(`/projects.json?limit=${limit}&offset=${offset}`, "GET").then(
        ({ total_count, projects }) => [
          total_count,
          projects.map(
            (proj) =>
              new RedmineProject(this, {
                ...proj,
              })
          ),
        ]
      );

      return req(offset + limit, limit, totalCount, accumulator.concat(result));
    };

    return req();
  }

  getTimeEntryActivities(): Promise<{
    time_entry_activities: TimeEntryActivity[];
  }> {
    if (this.timeEntryActivities) {
      return Promise.resolve({
        time_entry_activities: this.timeEntryActivities,
      });
    }
    return this.doRequest<{
      time_entry_activities: TimeEntryActivity[];
    }>(`/enumerations/time_entry_activities.json`, "GET").then((response) => {
      if (response) {
        this.timeEntryActivities = response.time_entry_activities;
      }

      return response;
    });
  }

  addTimeEntry(
    issueId: number,
    activityId: number,
    hours: string,
    message: string
  ): Promise<unknown> {
    return this.doRequest<{ time_entry: TimeEntry }>(
      `/time_entries.json`,
      "POST",
      Buffer.from(
        JSON.stringify({
          time_entry: <TimeEntry>{
            issue_id: issueId,
            activity_id: activityId,
            hours,
            comments: message,
          },
        })
      )
    );
  }

  /**
   * Returns promise, that resolves to an issue
   * @param issueId ID of issue
   */
  getIssueById(issueId: number): Promise<{ issue: Issue }> {
    return this.doRequest(`/issues/${issueId}.json`, "GET");
  }

  /**
   * Returns promise, that resolves, when issue status is set
   */
  setIssueStatus(issue: Issue, statusId: number): Promise<unknown> {
    return this.doRequest<{ issue: Issue }>(
      `/issues/${issue.id}.json`,
      "PUT",
      Buffer.from(
        JSON.stringify({
          issue: {
            status_id: statusId,
          },
        }),
        "utf8"
      )
    );
  }

  issueStatuses: { issue_statuses: RedmineIssueStatus[] } | null = null;

  /**
   * Returns promise, that resolves to list of issue statuses in provided redmine server
   */
  getIssueStatuses(): Promise<{ issue_statuses: RedmineIssueStatus[] }> {
    if (isNil(this.issueStatuses)) {
      return this.doRequest<{ issue_statuses: RedmineIssueStatus[] }>(
        "/issue_statuses.json",
        "GET"
      ).then((obj) => {
        if (obj) {
          // Shouldn't change much; cache it.
          this.issueStatuses = obj;
        }

        return obj;
      });
    } else {
      return Promise.resolve(this.issueStatuses);
    }
  }

  async getIssueStatusesTyped(): Promise<IssueStatus[]> {
    const statuses = await this.getIssueStatuses();
    return statuses.issue_statuses.map((s) => new IssueStatus(s.id, s.name));
  }
  async getMemberships(projectId: number): Promise<Membership[]> {
    const membershipsResponse = await this.doRequest<{
      memberships: RedmineMembership[];
    }>(`/projects/${projectId}/memberships.json`, "GET");

    return membershipsResponse.memberships.map((m) =>
      "user" in m
        ? new Membership(m.user.id, m.user.name)
        : new Membership(m.group.id, m.group.name, false)
    );
  }
  async applyQuickUpdate(quickUpdate: QuickUpdate): Promise<QuickUpdateResult> {
    await this.doRequest<void>(
      `/issues/${quickUpdate.issueId}.json`,
      "PUT",
      Buffer.from(
        JSON.stringify({
          issue: {
            status_id: quickUpdate.status.statusId,
            assigned_to_id: quickUpdate.assignee.id,
            notes: quickUpdate.message,
          },
        }),
        "utf8"
      )
    );
    const issueRequest = await this.getIssueById(quickUpdate.issueId);
    const issue = issueRequest.issue;
    const updateResult = new QuickUpdateResult();
    if (issue.assigned_to.id !== quickUpdate.assignee.id) {
      updateResult.addDifference("Couldn't assign user");
    }
    if (issue.status.id !== quickUpdate.status.statusId) {
      updateResult.addDifference("Couldn't update status");
    }
    return updateResult;
  }

  /**
   * Returns promise, that resolves to list of issues assigned to api key owner
   */
  getIssuesAssignedToMe(): Promise<{ issues: Issue[] }> {
    return this.doRequest<{ issues: Issue[] }>(
      "/issues.json?status_id=open&assigned_to_id=me",
      "GET"
    );
  }

  /**
   * Returns promise, that resolves to list of open issues for project
   */
  getOpenIssuesForProject(
    project_id: number | string,
    include_subproject = true
  ): Promise<{ issues: Issue[] }> {
    if (include_subproject) {
      return this.doRequest<{ issues: Issue[] }>(
        `/issues.json?status_id=open&project_id=${project_id}&subproject_id=!*`,
        "GET"
      );
    } else {
      return this.doRequest<{ issues: Issue[] }>(
        `/issues.json?status_id=open&project_id=${project_id}`,
        "GET"
      );
    }
  }

  compare(other: RedmineServer) {
    return (
      this.options.address === other.options.address &&
      this.options.key === other.options.key &&
      this.options.rejectUnauthorized === other.options.rejectUnauthorized &&
      isEqual(this.options.additionalHeaders, other.options.additionalHeaders)
    );
  }
}
